fix(orderbook): reduce invalid states and improve sync reliability

shamardy opened the pull request:
Better to be reviewed commit by commit, no need to review the first 2 commits https://github.com/KomodoPlatform/komodo-defi-framework/pull/2626/commits/a73a71429b0638da9fb3fce6978a55309bbafa9e and https://github.com/KomodoPlatform/komodo-defi-framework/pull/2626/commits/e3ad3ca6be972e1fbd40ab8e83fb08b6e9b95cb5 as they were the ones reverted in https://github.com/KomodoPlatform/komodo-defi-framework/pull/2631 and reapplied here. PR description to follow...

shamardy committed:
deps: add libp2p-allow-block-list

shamardy committed:
p2p: integrate allow/block list; add temp-ban/unban + expiry

shamardy committed:
p2p: temp-ban on orderbook SyncFailure (1h)

shamardy committed:
lp_network: add 120s per‑peer grace before temp‑banning on orderbook SyncFailure to prevent bans during initial sync

shamardy committed:
fix(ordermatch): safer sync + role/cause-aware bans; tighten keep-alive

- Sync: request diffs from our local “from” roots (SyncPlan), reducing FullTrie
  fallbacks and false mismatches; apply diffs tentatively and revert on mismatch.
- Bans: add cause/role-aware temp-ban with per-peer grace (120s, TTL 600s,
  temp-ban 20m). Relays do not ban Unavailable light nodes; light nodes ban
  Unavailable relays. Remote role detected via relay mesh query.
- Treat Ok(None) on SyncPubkeyOrderbookState as InvalidOrIncomplete (not
  Unavailable).
- Keep-alive: avoid creating pubkey state on stale messages (read-only pre-scan).
- Logging: single structured decision log for SyncFailure ban action.

Risk/rollout: minor policy tuning; no protocol changes. Verify behavior under
lossy links and observe reduced false bans.

shamardy committed:
p2p: track ban reasons; auto-unban connectivity to restore mesh

shamardy committed:
fix(p2p): use libp2p validation‑based forwarding
- Route message forwarding through libp2p’s validation path instead of manual propagation
- This achieves faster gossip cache cleanup while avoiding forwarding malformed/invalid messages as before

shamardy committed:
add a todo note about banning for other reasons like invalid topics among other things

shamardy committed:
add StaleKeepAlive.delay field to debug the logs for how late are the keepalive messages

shamardy committed:
docs(ordermatch,gossipsub): KA duplicates investigation; TODOs only

- Observed repeated PubkeyKeepAlive messages; root cause not confirmed yet.
  Hypotheses include seq no. based MessageId dedup gaps and equal-timestamp behavior.
- Added investigation notes and TODOs: considered content‑derived
  MessageId.
- Deferred code changes until we gather more evidence to avoid risky churn.

No functional changes; comments and guidance only.

shamardy committed:
feat(ordermatch,p2p): add optional `expected_roots` to `SyncPubkeyOrderbookState`
- Prepares for exact from→to trie diffs, enabling strict landing
  on expected roots (security/consistency hardening) while maintaining
  backward compatibility.

 Behavioral impact:
 - No functional change in this commit, purely protocol scaffolding.

shamardy committed:
feat(ordermatch,p2p): send `expected_roots` in `SyncPubkeyOrderbookState`
- Plumb `expected_roots` into the sync handler rejects unsound requests

shamardy committed:
feat(ordermatch): target exact to-root via from_history with temp FullTrie fallback

shamardy committed:
docs(ordermatch): add TODOs v2 sync tests

mariocynicys commented:
regarding [9bd5e53](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2626/commits/9bd5e532c87b283f99c8bb6667c8a1841bc1ae41)
i kinda want us to revert back to some set version in the CI & toolchain file and upgrade to the latest version every month or so (and maybe create a reminder bot for that). instead of using `stable` release.

i think @dimxy also did some clippy fixes in one of his PRs.
and i avoid doing these fixes (even though they are annoying AF in the terminal) mid other PRs to avoid conflicts if somebody else happens to fix them too.

but clippy fixes will usually create conflicts anyway even if done by a single person in an isolated PR :/, albeit less messy.

shamardy committed:
Revert "Revert "fix(ordermatch): ignore loop-back; clear on null root; reject stale keep-alives (#2580)""

This reverts commit fa3fc5dae0af15b9ea7ba5bfaf66cacdb26596da.

shamardy committed:
Revert "Revert "fix(orderbook): validate roots before commit (#2605)""

This reverts commit 9a76d11c2f505d3a325f16aa77fcfbec2d78ec40.

shamardy commented:
## Related: `test_order_cancellation_received_before_creation` ignored

The test `test_order_cancellation_received_before_creation` in `orderbook_sync_tests.rs` has been temporarily disabled with `#[ignore]` in PR #2707.

This test was failing intermittently with:
```
Timeout expired waiting for a log condition
```

The test appears to be validating order cancellation timing behavior that's related to the orderbook sync improvements in this PR. When this PR is ready, the test should either be:
1. Re-enabled if the improvements fix the timing issue
2. Rewritten to better test the new sync behavior
3. Deleted if it's no longer relevant

The `#[ignore]` comment points to this PR for tracking.
